velox/exec/Strings.cpp

```diff
@@ -19,40 +19,67 @@
 namespace facebook::velox::aggregate::prestosql {
 
 StringView Strings::append(StringView value, HashStringAllocator& allocator) {
+  // A string in Strings needs 8 bytes below itself for the last 8
+  // bytes of the previous string, which copied to the next string
+  // when setting a continued pointer. It also needs 8 bytes at its
+  // tail for the same continue pointer to the next.
+  constexpr int32_t kOverhead =
+      HashStringAllocator::Header::kContinuedPtrSize + 8;
   VELOX_DCHECK(!value.isInline());
 
-  maxStringSize = std::max(maxStringSize, value.size());
+  maxStringSize = std::max<int32_t>(maxStringSize, value.size());
 
   // Request sufficient amount of memory to store the whole string
-  // (value.size()) and allow some memory left for bookkeeping (header + link
-  // to next block).
-  const int32_t requiredBytes =
-      value.size() + HashStringAllocator::Header::kContinuedPtrSize + 8;
-
+  // (value.size()) and allow some memory left for bookkeeping (8 last from
+  // previous block + link to next block).
+  const int32_t requiredBytes = value.size() + kOverhead;
+  int32_t tail = 0;
   ByteStream stream(&allocator);
   if (firstBlock == nullptr) {
     // Allocate first block.
     currentBlock = allocator.newWrite(stream, requiredBytes);
     firstBlock = currentBlock.header;
+    auto minSize = HashStringAllocator::freeListSizes()[0];
+    tail = requiredBytes < minSize ? minSize - requiredBytes : 0;
   } else {
     allocator.extendWrite(currentBlock, stream);
+    tail = stream.ranges().back().size;
   }
 
   // Check if there is enough space left.
-  if (stream.ranges().back().size < requiredBytes) {
+  int32_t available = stream.ranges().back().size;
+  if (available <
+      value.size() + HashStringAllocator::Header::kContinuedPtrSize) {
     // Not enough space. Allocate new block.
     ByteRange newRange;
-    allocator.newContiguousRange(requiredBytes, &newRange);
+    auto sizes = HashStringAllocator::freeListSizes();
+    int32_t targetBytes = kOverhead +
+        (maxStringSize < 200      ? maxStringSize * 4
+             : value.size() < 200 ? value.size() * 4
+                                  : value.size());
+    int32_t roundedUpBytes = targetBytes < sizes[0] ? sizes[0]
+        : targetBytes < sizes[1]                    ? sizes[1]
+                                                    : targetBytes;
```

Comment:
I want to check my understanding here, if the string would fall in either of the first two buckets, you round the size up so you look in the next bucket, and since you're requesting the smallest possible chunk in whichever bucket searching the free list in this case amounts to just checking if the free list is empty.

why do you do the same for the second bucket, but no futher? why not just the first bucket?